### ISSUE TITLE
Updates API / OTA : Android disk usage grows with each OTA update

### ISSUE THREAD

**Issue by bonham000** [expo/expo#2896]
### Environment

Any Expo app.

### Steps to Reproduce

1. Create an Expo app and publish it to the app stores.
2. Install the app on your device.
3. Publish OTA updates for your app.
4. View the app memory footprint on the device before and after the update.

### Expected Behavior

Previous (stale) JS bundles are cleared from the device after a new JS bundle is loaded over the air.

### Actual Behavior

All prior JS bundles remain on device, this causes the app's memory usage to grow over time, increasing with each OTA release. For instance, here is the progression of the data reported by Android for our staging app from a fresh install through 3 OTA updates (each update was the same code, just published again):

```
Fresh install:    32KB
1st OTA update: ~ 8.8MB
2nd OTA update: ~ 12.5MB
3rd OTA update: ~ 18.24MB
```

### Extra Info

I first [reported this yesterday on the Expo Forums](https://forums.expo.io/t/expo-apk-size-grows-with-each-ota-update/16593) and was recommended to open an issue here. Previously, our team had assumed Expo apps would keep only the current JS bundle saved on the device, but it appears all downloaded JS bundles remain saved on the device.

This creates a scenario where an app's memory footprint on a device grows over time, increasing with each OTA update. For JS bundles that are 1-5MB, this could result in app sizes which take up 100s of MB after many updates are pushed. Because of how easy it easy to publish updates OTA, this is a realistic possibility, but very problematic for users and developers if the app size gets to be this large. We first noticed this looking at our staging build (which we publish frequently) which was taking up 200-300MB in total size.

### Questions

- Our team has demonstrated this behavior on Android, but not yet on iOS. It's unclear to me right now if this behavior is the same on iOS or not. **Update:** I'm pretty sure this **does not** happen on iOS.
- We have not tested yet to see the result if we push new binaries to the App Stores. If pushing a new app binary and updating it clears the extra JS bundles, then this is reasonable to workaround for now because we expect to be building new binaries every 1-3 months or so to keep up with Expo SDK releases.
- It would be great if Expo could clear stale JS bundles after loading a new OTA update, or, provide some API where developers can opt in to removing these (similar to removing data from the `FileSystem` provided by Expo.
- It may be helpful to add more clear documentation about this behavior in the OTA updates section, otherwise people may push updates frequently without being aware of the potential impact to the app size.

***

Please let me know if I missed anything or if my understanding of Expo is incorrect!

**Comment by quinlanj** [comment]
Thanks for reporting this issue. We will look into it :)

**Comment by amsimoes** [comment]
@bonham000 What makes you pretty sure this doesn't happen in iOs?

Also, thanks a lot on reporting this. I think this must be on the documentation, and it should not be the default behaviour. What's the reason behind this anyways?

Cheers

**Comment by bonham000** [comment]
@amsimoes we've been watching our staging apps as we push updates on iOS and Android and Android demonstrates the behavior described above whereas iOS seems to maintain a constant memory footprint on the device when it downloads new OTA updates, so it seems like the data usage on the device does not grow with each update on iOS.

**Comment by amsimoes** [comment]
@bonham000 Thank you so much for this precious information!! I think it's a critical info for anyone wondering about using OTA on production...

Anyone from the Expo team care to comment? Can we change this behavior? @quinlanj

**Comment by amsimoes** [comment]
Bump.

@AdamJNavarro @quinlanj 

Any progress on this?
I think this is a major issue that should at least be mentioned on OTA docs page...

Thank you.

**Comment by ncailleux** [comment]
I've seeing this too with my application. Nothing really change on iOS, but on android each OTA makes the size bigger and bigger.
I hope that can be fixed in the future, OTA is a very nice feature, but not with this behaviour.
Thanks

**Comment by dingxiangfei2009** [comment]
cf. https://github.com/expo/expo/blob/2d45a567213722f0453d2151b5c551135d3036bb/android/expoview/src/main/java/host/exp/expoview/Exponent.java#L353

It does not clean them up, yet.

**Comment by dingxiangfei2009** [comment]
@amsimoes iOS seems to pile up the OTA bundles, too. If Expo decides that it should invalidate cache, it will clean it up, but the decision procedure is hardcoded to not invalidate any cache.

**Comment by esamelson** [comment]
reopening, a PR was merged that closes this issue but we haven't released the fix yet.

**Comment by ide** [comment]
The current system downloads updates to the OS's cache directory, which will get cleaned up if the OS deems it necessary. This means the actual problem is that updates can get removed by the OS. The growing disk usage is just a visible symptom but not an actually critical issue because it's transient (cache) disk usage, not persistent (documents) disk usage.

We are looking at working on a new Updates module for later this year that will be designed to track resources more closely and will address both issues in a fundamental way, rather than patching a fix on top.

**Comment by jhalborg** [comment]
Thanks @ide ! 👏

Could it perhaps be possible to hope for both, i.e. a patch now and a larger makeover of the Update module later?

**Comment by ide** [comment]
There is a PR out for a method to manually delete JS bundles. Fair warning: you can delete an app's latest copy of the bundle (i.e. it is dangerous) and the method will be undocumented, since we are putting our support towards the new Updates module for 2019.

**Comment by esamelson** [comment]
The method in the referenced PR (#4851) has now been released in `expo@34.0.4`. While it will not work in the expo client, it will work in SDK 34 builds made with `expo build:android`. Again, at this time we are not documenting or providing support for this method and do NOT recommend using it to delete the currently running OTA update bundle (which is very possible to do!) so use carefully. It should be possible and straightforward to delete older OTA updates from previous SDK versions and clear up some space on the device.

**Comment by esamelson** [comment]
Closing as the fix has been released. We are working on a new Updates module which will have this functionality built in from the start.
